File: fastpay-assistant-ml/fastpay_assistant/ml/entity_extractor.py
from __future__ import annotations

import re

from fastpay_assistant.types import ExtractedEntities

_AMOUNT_RE = re.compile(
    r"(?:rwf\s*)?(\d{1,3}(?:,\d{3})+|\d{4,}|\d+)(?:\s*(?:rwf|frw))?",
    re.I,
)
_ASSET_RE = re.compile(r"\b(usdt|btc|bitcoin|sol|solana)\b", re.I)
_TIMEFRAME_RE = re.compile(
    r"\b(today|this week|this month|last month|weekly|monthly|yearly|this year)\b",
    re.I,
)
_ACTION_RE = re.compile(r"\b(transfer|send|save|savings|loan|borrow)\b", re.I)

_ASSET_MAP = {
    "usdt": "USDT",
    "btc": "BTC",
    "bitcoin": "BTC",
    "sol": "SOL",
    "solana": "SOL",
}

_ACTION_MAP = {
    "transfer": "transfer",
    "send": "transfer",
    "save": "save",
    "savings": "save",
    "loan": "loan",
    "borrow": "loan",
}


def extract_entities(message: str) -> ExtractedEntities:
    amount: float | None = None
    for match in _AMOUNT_RE.finditer(message):
        raw = match.group(1).replace(",", "").replace(" ", "")
        try:
            value = float(raw)
        except ValueError:
            continue
        # skip tiny numbers that are likely years / levels unless currency nearby
        span = message[max(0, match.start() - 4) : match.end() + 4].lower()
        if value < 100 and "rwf" not in span and "frw" not in span:
            continue
        amount = value
        break

    asset = None
    am = _ASSET_RE.search(message)
    if am:
        asset = _ASSET_MAP[am.group(1).lower()]  # type: ignore[assignment]

    timeframe = None
    tm = _TIMEFRAME_RE.search(message)
    if tm:
        timeframe = tm.group(1).lower()

    action = None
    act = _ACTION_RE.search(message)
    if act:
        action = _ACTION_MAP[act.group(1).lower()]  # type: ignore[assignment]

    return ExtractedEntities(
        amount_rwf=amount,
        asset=asset,
        timeframe=timeframe,
        action=action,
    )

File: FastPay/lib/assistant/ml/entity-extractor.ts
import type { ExtractedEntities } from "../types";

const AMOUNT_RE =
  /(?:rwf\s*)?(\d{1,3}(?:,\d{3})+|\d{4,}|\d+)(?:\s*(?:rwf|frw))?/gi;
const ASSET_RE = /\b(usdt|btc|bitcoin|sol|solana)\b/i;
const TIMEFRAME_RE =
  /\b(today|this week|this month|last month|weekly|monthly|yearly|this year)\b/i;
const ACTION_RE = /\b(transfer|send|save|savings|loan|borrow)\b/i;

const ASSET_MAP: Record<string, ExtractedEntities["asset"]> = {
  usdt: "USDT",
  btc: "BTC",
  bitcoin: "BTC",
  sol: "SOL",
  solana: "SOL",
};

const ACTION_MAP: Record<string, ExtractedEntities["action"]> = {
  transfer: "transfer",
  send: "transfer",
  save: "save",
  savings: "save",
  loan: "loan",
  borrow: "loan",
};

export function extractEntities(message: string): ExtractedEntities {
  let amountRwf: number | undefined;

  for (const match of message.matchAll(AMOUNT_RE)) {
    const raw = match[1].replace(/,/g, "");
    const value = Number(raw);
    if (Number.isNaN(value)) {
      continue;
    }
    const span = message
      .slice(Math.max(0, match.index! - 4), match.index! + match[0].length + 4)
      .toLowerCase();
    if (value < 100 && !span.includes("rwf") && !span.includes("frw")) {
      continue;
    }
    amountRwf = value;
    break;
  }

  const assetMatch = message.match(ASSET_RE);
  const asset = assetMatch ? ASSET_MAP[assetMatch[1].toLowerCase()] : undefined;

  const timeframeMatch = message.match(TIMEFRAME_RE);
  const timeframe = timeframeMatch?.[1].toLowerCase();

  const actionMatch = message.match(ACTION_RE);
  const action = actionMatch
    ? ACTION_MAP[actionMatch[1].toLowerCase()]
    : undefined;

  return { amountRwf, asset, timeframe, action };
}
